File: .rss-tmp/parse.js
#!/usr/bin/env node
const fs = require("fs");
const path = require("path");

const FEED_NAMES = {
  cbc_politics: "CBC Politics",
  canadaland: "Canadaland",
  globe_opinion: "Globe and Mail Opinion",
  the_tyee: "The Tyee",
  walrus_politics: "The Walrus Politics",
  ipolitics: "iPolitics",
  breach_media: "Breach Media",
  pluralistic: "Pluralistic",
  disconnect: "Disconnect",
  michael_geist: "Michael Geist",
  policy_options: "Policy Options",
  priv_canada: "Privacy Commissioner of Canada",
  the_markup: "The Markup",
  eff: "EFF",
  betakit: "BetaKit",
  citation_needed: "Citation Needed",
  jwz: "JWZ",
  schneier: "Schneier on Security",
  joan_westenberg: "Joan Westenberg",
  ed_zitron: "Ed Zitron",
  convivial: "The Convivial Society",
  contraptions: "Contraptions",
  fourzerofour: "404 Media",
  platformer: "Platformer",
  garbage_day: "Garbage Day",
  tech_policy_press: "Tech Policy Press",
  techdirt: "Techdirt",
  robin_sloan: "Robin Sloan",
  ted_gioia: "Ted Gioia",
  dada_drummer: "Dada Drummer",
  aquarium_drunkard: "Aquarium Drunkard",
  interdependence: "Interdependence",
  metalabel: "Metalabel",
  new_models: "New Models",
  kneeling_bus: "Kneeling Bus",
  real_life: "Real Life",
  nplusone: "n+1",
  unpopular_front: "Unpopular Front",
  yancey_strickler: "Yancey Strickler",
};

const NOW = new Date();
const CUTOFF = new Date(NOW.getTime() - 48 * 3600 * 1000);

function stripHtml(s) {
  if (!s) return "";
  return s
    .replace(/<!\[CDATA\[/g, "")
    .replace(/\]\]>/g, "")
    .replace(/<[^>]+>/g, " ")
    .replace(/&nbsp;/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&#8217;/g, "'")
    .replace(/&#8216;/g, "'")
    .replace(/&#8220;/g, '"')
    .replace(/&#8221;/g, '"')
    .replace(/&#8211;/g, "-")
    .replace(/&#8212;/g, "—")
    .replace(/\s+/g, " ")
    .trim();
}

function parseDate(s) {
  if (!s) return null;
  s = s.trim();
  const d = new Date(s);
  if (isNaN(d.getTime())) return null;
  return d;
}

function extract(block, tag) {
  // Matches <tag ...>...</tag> non-greedy across newlines
  const re = new RegExp(`<${tag}(?:\\s+[^>]*)?>([\\s\\S]*?)</${tag}>`, "i");
  const m = block.match(re);
  return m ? m[1] : "";
}

function extractAttr(block, tag, attr) {
  const re = new RegExp(`<${tag}\\s+[^>]*?${attr}=["']([^"']+)["']`, "i");
  const m = block.match(re);
  return m ? m[1] : "";
}

function parseFeed(xml, key) {
  const items = [];
  // RSS 2.0: <item>...</item>
  const itemRe = /<item(?:\s[^>]*)?>([\s\S]*?)<\/item>/gi;
  let m;
  while ((m = itemRe.exec(xml)) !== null) {
    const block = m[1];
    const title = stripHtml(extract(block, "title"));
    let link = stripHtml(extract(block, "link"));
    if (!link) {
      link = extractAttr(block, "link", "href");
    }
    const dateStr =
      extract(block, "pubDate") ||
      extract(block, "dc:date") ||
      extract(block, "published") ||
      extract(block, "updated");
    const desc = extract(block, "description");
    const content = extract(block, "content:encoded") || extract(block, "content");
    const d = parseDate(stripHtml(dateStr));
    if (!d || d < CUTOFF) continue;
    items.push({
      title,
      link,
      date: d.toISOString(),
      summary: stripHtml(content || desc).slice(0, 700),
      feed: FEED_NAMES[key] || key,
    });
  }
  // Atom: <entry>...</entry>
  const entryRe = /<entry(?:\s[^>]*)?>([\s\S]*?)<\/entry>/gi;
  while ((m = entryRe.exec(xml)) !== null) {
    const block = m[1];
    const title = stripHtml(extract(block, "title"));
    let link = extractAttr(block, "link", "href");
    if (!link) link = stripHtml(extract(block, "link"));
    const dateStr =
      extract(block, "published") ||
      extract(block, "updated") ||
      extract(block, "pubDate");
    const summary = extract(block, "summary");
    const content = extract(block, "content");
    const d = parseDate(stripHtml(dateStr));
    if (!d || d < CUTOFF) continue;
    items.push({
      title,
      link,
      date: d.toISOString(),
      summary: stripHtml(content || summary).slice(0, 700),
      feed: FEED_NAMES[key] || key,
    });
  }
  return items;
}

const base = path.dirname(path.resolve(__filename));
const files = fs.readdirSync(base).filter((f) => f.endsWith(".xml") && f !== "test.xml");
const allItems = [];
const counts = {};
for (const f of files) {
  const key = f.replace(".xml", "");
  const xml = fs.readFileSync(path.join(base, f), "utf8");
  if (xml.length < 100) {
    counts[FEED_NAMES[key] || key] = 0;
    continue;
  }
  const items = parseFeed(xml, key);
  counts[FEED_NAMES[key] || key] = items.length;
  allItems.push(...items);
}
allItems.sort((a, b) => b.date.localeCompare(a.date));
fs.writeFileSync(
  path.join(base, "items.json"),
  JSON.stringify({ items: allItems, counts, cutoff: CUTOFF.toISOString(), now: NOW.toISOString() }, null, 2)
);
console.log(`Total items (last 48h): ${allItems.length}`);
const withItems = Object.entries(counts).filter(([, v]) => v > 0);
console.log(`Feeds with items: ${withItems.length}/${Object.keys(counts).length}`);
for (const [name, count] of withItems.sort((a, b) => b[1] - a[1])) {
  console.log(`  ${name}: ${count}`);
}
const zero = Object.entries(counts).filter(([, v]) => v === 0).map(([n]) => n);
if (zero.length) console.log(`Zero items: ${zero.join(", ")}`);

File: .rss-tmp/parse.py
#!/usr/bin/env python3
"""Parse RSS/Atom feeds, extract last 48h entries."""
import os
import re
import json
import glob
from datetime import datetime, timezone, timedelta
from email.utils import parsedate_to_datetime
import xml.etree.ElementTree as ET

FEED_NAMES = {
    "cbc_politics": "CBC Politics",
    "canadaland": "Canadaland",
    "globe_opinion": "Globe and Mail Opinion",
    "the_tyee": "The Tyee",
    "walrus_politics": "The Walrus Politics",
    "ipolitics": "iPolitics",
    "breach_media": "Breach Media",
    "pluralistic": "Pluralistic",
    "disconnect": "Disconnect",
    "michael_geist": "Michael Geist",
    "policy_options": "Policy Options",
    "priv_canada": "Privacy Commissioner of Canada",
    "the_markup": "The Markup",
    "eff": "EFF",
    "betakit": "BetaKit",
    "citation_needed": "Citation Needed",
    "jwz": "JWZ",
    "schneier": "Schneier on Security",
    "joan_westenberg": "Joan Westenberg",
    "ed_zitron": "Ed Zitron",
    "convivial": "The Convivial Society",
    "contraptions": "Contraptions",
    "fourzerofour": "404 Media",
    "platformer": "Platformer",
    "garbage_day": "Garbage Day",
    "tech_policy_press": "Tech Policy Press",
    "techdirt": "Techdirt",
    "robin_sloan": "Robin Sloan",
    "ted_gioia": "Ted Gioia",
    "dada_drummer": "Dada Drummer",
    "aquarium_drunkard": "Aquarium Drunkard",
    "interdependence": "Interdependence",
    "metalabel": "Metalabel",
    "new_models": "New Models",
    "kneeling_bus": "Kneeling Bus",
    "real_life": "Real Life",
    "nplusone": "n+1",
    "unpopular_front": "Unpopular Front",
    "yancey_strickler": "Yancey Strickler",
}

NOW = datetime.now(timezone.utc)
CUTOFF = NOW - timedelta(hours=48)

NS = {
    "atom": "http://www.w3.org/2005/Atom",
    "dc": "http://purl.org/dc/elements/1.1/",
    "content": "http://purl.org/rss/1.0/modules/content/",
}


def parse_date(s):
    if not s:
        return None
    s = s.strip()
    try:
        dt = parsedate_to_datetime(s)
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt
    except Exception:
        pass
    for fmt in ("%Y-%m-%dT%H:%M:%S%z", "%Y-%m-%dT%H:%M:%SZ", "%Y-%m-%dT%H:%M:%S.%f%z", "%Y-%m-%dT%H:%M:%S.%fZ"):
        try:
            s2 = s.replace("Z", "+0000")
            dt = datetime.strptime(s2, fmt)
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            return dt
        except Exception:
            pass
    return None


def strip_html(s):
    if not s:
        return ""
    s = re.sub(r"<[^>]+>", " ", s)
    s = re.sub(r"&nbsp;", " ", s)
    s = re.sub(r"&amp;", "&", s)
    s = re.sub(r"&quot;", '"', s)
    s = re.sub(r"&#39;", "'", s)
    s = re.sub(r"&lt;", "<", s)
    s = re.sub(r"&gt;", ">", s)
    s = re.sub(r"\s+", " ", s)
    return s.strip()


def parse_feed(path, key):
    try:
        with open(path, "rb") as f:
            data = f.read()
        if len(data) < 100:
            return []
        root = ET.fromstring(data)
    except Exception as e:
        return []

    items = []
    # RSS 2.0
    for item in root.iter("item"):
        title = (item.findtext("title") or "").strip()
        link = (item.findtext("link") or "").strip()
        date_str = item.findtext("pubDate") or item.findtext("{http://purl.org/dc/elements/1.1/}date") or ""
        desc = item.findtext("description") or ""
        content = item.findtext("{http://purl.org/rss/1.0/modules/content/}encoded") or ""
        dt = parse_date(date_str)
        if not dt or dt < CUTOFF:
            continue
        items.append({
            "title": strip_html(title),
            "link": link,
            "date": dt.isoformat(),
            "summary": strip_html(content or desc)[:600],
            "feed": FEED_NAMES.get(key, key),
        })
    # Atom
    for entry in root.iter("{http://www.w3.org/2005/Atom}entry"):
        title = (entry.findtext("{http://www.w3.org/2005/Atom}title") or "").strip()
        link_el = entry.find("{http://www.w3.org/2005/Atom}link")
        link = link_el.get("href") if link_el is not None else ""
        date_str = entry.findtext("{http://www.w3.org/2005/Atom}published") or entry.findtext("{http://www.w3.org/2005/Atom}updated") or ""
        summary = entry.findtext("{http://www.w3.org/2005/Atom}summary") or ""
        content = entry.findtext("{http://www.w3.org/2005/Atom}content") or ""
        dt = parse_date(date_str)
        if not dt or dt < CUTOFF:
            continue
        items.append({
            "title": strip_html(title),
            "link": link,
            "date": dt.isoformat(),
            "summary": strip_html(content or summary)[:600],
            "feed": FEED_NAMES.get(key, key),
        })
    return items


def main():
    base = os.path.dirname(os.path.abspath(__file__))
    all_items = []
    feed_counts = {}
    for path in sorted(glob.glob(os.path.join(base, "*.xml"))):
        key = os.path.basename(path).replace(".xml", "")
        if key == "test":
            continue
        items = parse_feed(path, key)
        feed_counts[FEED_NAMES.get(key, key)] = len(items)
        all_items.extend(items)
    all_items.sort(key=lambda x: x["date"], reverse=True)
    with open(os.path.join(base, "items.json"), "w") as f:
        json.dump({"items": all_items, "counts": feed_counts, "cutoff": CUTOFF.isoformat(), "now": NOW.isoformat()}, f, indent=2)
    print(f"Total items: {len(all_items)}")
    print(f"Feeds with items: {sum(1 for v in feed_counts.values() if v > 0)}")
    print(f"Feeds with zero items: {sum(1 for v in feed_counts.values() if v == 0)}")
    for fname, count in sorted(feed_counts.items(), key=lambda x: -x[1]):
        if count > 0:
            print(f"  {fname}: {count}")


if __name__ == "__main__":
    main()
